auth/api/iam/s2s_vptoken.go

```diff
@@ -19,49 +19,157 @@
-		// TODO: right error response
-		return "", errors.New("missing required parameters")
-	}
-	// TODO: https://github.com/nuts-foundation/nuts-node/issues/2418
-	// TODO: verify parameters
-	return scope, nil
+	// Check signatures and VC issuer trust
+	issueTime := time.Now()
+	_, err = s.vcr.Verifier().VerifyVP(vp, true, false, &issueTime)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "verifiable presentation is invalid",
+			InternalError: err,
+		}
+	}
+
+	// Validate the presentation submission:
+	// 1. Resolve presentation definition for the requested scope
+	// 2. Take VCs mapped by the presentation submission
+	// 3. Match the VCs from (2) to the definition from (1). This should yield the same list of credentials.
+	//    This actually creates a new submission.
+	presentationDefinition := s.auth.PresentationDefinitions().ByScope(scope)
+	if presentationDefinition == nil {
+		return nil, OAuth2Error{
+			Code:        InvalidScope,
+			Description: fmt.Sprintf("unsupported scope for presentation exchange: %s", scope),
+		}
+	}
+	submissionCredentials, err := submission.Credentials(vp)
+	if err != nil {
+		return nil, err
+	}
+	// Match the credentials from the presentation submission with the presentation definition we got through the scope
+	_, matchedCredentials, err := presentationDefinition.Match(submissionCredentials)
+	if err != nil {
+		return nil, OAuth2Error{
+			Code:          InvalidRequest,
+			Description:   "presentation_submission parameter is invalid: re-evaluation of the credentials against the presentation definition fails",
+			InternalError: err,
+		}
+	}
+	submissionCredentialsJSON, _ := json.Marshal(submissionCredentials)
+	matchedCredentialsJSON, _ := json.Marshal(matchedCredentials)
+	if !bytes.Equal(submissionCredentialsJSON, matchedCredentialsJSON) {
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "presentation_submission parameter is invalid: re-evaluation of the credentials against the presentation definition yields different credentials (bug or attempted hack)",
+		}
+	}
+	// All OK, allow access
+	response, err := s.createAccessToken(issuer, issueTime, vp, scope)
+	if err != nil {
+		return nil, err
+	}
+	return HandleTokenRequest200JSONResponse(*response), nil
 }
 
-func (s serviceToService) handleAuthzRequest(_ map[string]string, _ *Session) (*authzResponse, error) {
-	// Protocol does not support authorization code flow
-	return nil, nil
+func (s Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presentation vc.VerifiablePresentation, scope string) (*TokenResponse, error) {
+	accessToken := AccessToken{
+		Token:        generateCode(),
+		Issuer:       issuer.String(),
+		Expiration:   issueTime.Add(accessTokenValidity),
+		Presentation: presentation,
+	}
+	err := s.storageInstance.GetSessionDatabase().GetStore(accessTokenValidity, issuer.String(), "accesstoken").Put(accessToken.Token, accessToken)
```

Comment:
maybe add `s2s` as first key so we can debug how many s2s keys are active?

Reply:
Yeah, I prefixed the OpenID keys with `openid`, so `s2s` is in order here I think